bash: Add tests.withChecks passthru to run tests

infinisil opened the pull request:
Allows running bash tests with
```
nix-build -A bash.tests.withChecks
```

This notably fails if a test fails, which is not a given with bash tests!

Note also that we can't run the tests by default (without other hacks) because the tests depend on a bunch of packages with bash as a dependency.

Note also that we had to disable some tests for now. In the future we could figure out how to make those work too.

CC @balsoft @trofi 

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

balsoft commented on pkgs/shells/bash/fail-tests.patch:
```diff
@@ -0,0 +1,19 @@
++    echo $x
++    if ! sh $x; then
++      echo "Test $x failed"
++      #exit 1
```
Uncomment this before merging

balsoft reviewed (changes requested):
Actually, just noticed we missed a thing

infinisil committed:
bash: Add tests.withChecks passthru to run tests

Co-Authored-By: Alexander Bantyev <alexander.bantyev@tweag.io>

infinisil committed:
bash: Add myself to maintainers

trofi commented:
> @trofi [...] would you also be okay with adding yourself?

No, I'll pass. I only have time for small drive-by changes.

balsoft reviewed (approved):
Now looks good to me!

infinisil commented:
10 more tests were failing on Darwin, so I now just marked the test derivation as broken on Darwin because I don't want to bother trying to make them work or marking them as broken individually.

If we spend more time on this, let's focus on the Linux tests first.

infinisil commented:
@balsoft In addition to addressing your comment (and other improvements to the patch), I removed this:

```bash
# Fail if diff fails at the end of each test script
for check in tests/run-*; do
  # Append "exit $?" at the very end of the file
  sed -i '$aexit $?' "$check"
done
```

Everything works (or fails when expected) as it should, I guess that part isn't necessary after all. As a reminder, all the test files look like 
```bash
${THIS_SH} ./arith.tests > ${BASH_TSTOUT} 2>&1
diff ${BASH_TSTOUT} arith.right && rm -f ${BASH_TSTOUT}
```

If the `diff` fails (aka there is a diff), this script exits with 1 and doesn't remove the test file.
